ops/model.py

```diff
@@ -996,10 +997,15 @@ def relation_get(self, relation_id, member_name, is_app):
         if not isinstance(is_app, bool):
             raise TypeError('is_app parameter to relation_get must be a boolean')
 
+        if is_app and not JujuVersion.from_environ().has_app_data():
+            return {}
```

Comment:
I'm hesitant on this one. Depending on how they're using application data, they may *need* it to be there. So I think I'd rather have a different error that is clearer about 'this isn't supported by Juju' and charms can catch that exception and deal with it.
Silently returning {} is dangerous, IMO.

Also, the bugs in 2.7.* were that you couldn't read your *own* application data. Reading the application data for a related application works fine. (app A related to app B, A/0 can read B's application data, and B/0 can read A's application data, the bug was that the leader A/0 couldn't read A's application data.)

Which means that this would preclude charms from using relation application data even where Juju *did* support it.